enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -248,6 +247,73 @@ type ControlPlaneMachineSetStatus struct {
     // +optional
     Conditions []metav1.Condition `json:"conditions,omitempty"`
 }
+
+// FailureDomain represents the different configurations required to spread Machines
+// across failure domains on different platforms.
+// +union
+type FailureDomain struct {
+  // Platform identifies the platform for which the FailureDomain represents
+  // +kubebuilder:validation:Enum:="aws";"azure";"gcp";"openstack"
+  // +unionDiscriminator
+  // +optional
+  Platform configv1.PlatformType `json:platform,omitempty`
+
+  // AWS configures failure domain information for the AWS platform
+  // +optional
+  AWS *AWSFailureDomain `json:aws,omitempty`
```

Comment:
Enumerating all the things... Is there really no other way?

Incidentally, I do really like the idea of a structure containing infrastructure defaults specific to an AZ. I suspect you'll soon want to reference these from MachineSet/Machine.

Reply:
> Enumerating all the things... Is there really no other way?

The alternative is to have a list of raw extension and allow users to put any yaml in here that we want, and then writing something to overlay that over the providerSpec that they provide us.

I think it's preferable to have this structured and control what we will allow them to vary per failure domain.

The alternative is that information such as this is move centrally out of the Machine provider specs (as is done in CAPI), but then that's a major change to MAPI which we can't afford to make right now.